src/ripple/nodestore/ShardValidation.md

```diff
@@ -0,0 +1,130 @@
+
+We employ a simple heuristic for determining whether the application has stored a ledger later than the last ledger of the downloaded shard:
+
+```C++
+// We use the presence (or absense) of the validated
+// ledger as a heuristic for determining whether or
+// not we have stored a ledger that comes after the
+// last ledger in this shard. A later ledger must
+// be present in order to reliably retrieve the hash
+// of the shard's last ledger.
+if (app_.getLedgerMaster().getValidatedLedger())
+{
+    auto const hash = app_.getLedgerMaster().getHashBySeq(
+        lastLedgerSeq(shardIndex));
+
+    .
+    .
+    .
+}
+```
+
+The `getHashBySeq` function will be invoked only when a call to `LedgerMaster::getValidatedLedger` returns a validated ledger, rather than a `nullptr`. Otherwise validation of the shard will be deferred.
+
+### Retries
+
+#### Retry Limit
+
+If the server must defer shard validation, the software will initiate a timer that upon expiration, will re-attempt confirming the last ledger hash. We place an upper limit on the number of attempts the server makes to achieve this confirmation. When the maximum number of attempts has been reached, validation of the shard will fail, resulting in the removal of the shard. An attempt counts toward the limit only when we are able to get a validated ledger (implying a current view of the network), but are unable to retrieve the last ledger hash. Retries that occur because no validated ledger was available are not counted.
+
+#### ShardInfo
+
+The `DatabaseShardImp` class stores a container of `ShardInfo` structs, each of which contains information pertaining to a shard held by the server. These structs will be used during shard import to store the last ledger hash (when available) and to track the number of hash confirmation attempts that have been made.
+
+```C++
+struct ShardInfo
+{
+    .
+    .
+    .
+
+    // Used to limit the number of times we attempt
+    // to retrieve a shard's last ledger hash, when
+    // the hash should have been found. See
+    // scheduleFinalizeShard(). Once this limit has
+    // been exceeded, the shard has failed and is
+    // removed.
+    bool
+    attemptHashConfirmation()
+    {
+        if (lastLedgerHashAttempts + 1 <= maxLastLedgerHashAttempts)
+        {
+            ++lastLedgerHashAttempts;
+            return true;
+        }
+
+        return false;
+    }
+
+    // This variable is used during the validation
+    // of imported shards and must match the
+    // imported shard's last ledger hash.
+    uint256 lastLedgerHash;
+
+    // The number of times we've attempted to
+    // confirm this shard's last ledger hash.
+    uint16_t lastLedgerHashAttempts;
+
+    // The upper limit on attempts to confirm
+    // the shard's last ledger hash.
+    static const uint8_t maxLastLedgerHashAttempts = 5;
```

Comment:
Same proposition about std::uint16_t and std::uint8_t.

Reply:
This code snippet has been removed.